source/kernel/src/services/sdmmc.rs

```diff
@@ -0,0 +1,489 @@
+/// | Reserved              | `[23:20]`   |
+/// | Manufacturing date    | `[19:8]`    |
+/// | CRC7 checksum         | `[7:1]`     |
+/// | Not used, always 1    | `[0:0]`     |
+pub struct CardIdentification(u128);
+
+/// Published RCA in R6 response format
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub struct RelativeCardAddress(u32);
+
+impl SdCardClient {
+    /// Obtain an `SdCardClient`
+    ///
+    /// If the [`SdmmcService`] hasn't been registered yet, we will retry until it
+    /// has been registered.
+    pub async fn from_registry(
+        kernel: &'static Kernel,
+    ) -> Result<Self, registry::ConnectError<SdmmcService>> {
+        let handle = kernel.registry().connect::<SdmmcService>(()).await?;
+
+        Ok(Self {
+            handle,
+            reply: Reusable::new_async().await,
+        })
+    }
+
+    /// Obtain an `SdCardClient`
+    ///
+    /// Does NOT attempt to get an [`SdmmcService`] handle more than once.
+    ///
+    /// Prefer [`SdCardClient::from_registry`] unless you will not be spawning one
+    /// around the same time as obtaining a client.
+    pub async fn from_registry_no_retry(
+        kernel: &'static Kernel,
+    ) -> Result<Self, registry::ConnectError<SdmmcService>> {
+        let handle = kernel.registry().try_connect::<SdmmcService>(()).await?;
+
+        Ok(Self {
+            handle,
+            reply: Reusable::new_async().await,
+        })
+    }
+
+    async fn cmd(&mut self, cmd: Command) -> Result<Response, Error> {
+        self.handle
+            .request_oneshot(cmd, &self.reply)
+            .await
+            .map_err(|error| {
+                tracing::warn!(?error, "failed to send request to SD/MMC service");
+                Error::Other // TODO
+            })
+            .and_then(|resp| resp.body)
+    }
+
+    /// Reset the card
+    pub async fn reset(&mut self) -> Result<(), Error> {
+        self.cmd(Command::default()).await.map(|_| ())
+    }
+
+    /// Initialize the card
+    pub async fn initialize(&mut self) -> Result<CardType, Error> {
+        /// Request switch to 1.8V
+        #[allow(dead_code)]
+        const OCR_S18R: u32 = 0x1000000;
+        /// Host supports high capacity
+        const OCR_HCS: u32 = 0x40000000;
+        /// Card has finished power up routine if bit is high
+        const OCR_NBUSY: u32 = 0x80000000;
+        /// Valid bits for voltage setting
+        const OCR_VOLTAGE_MASK: u32 = 0x007FFF80;
+
+        let mut card_type = CardType::SD1;
+
+        match self
+            .cmd(Command {
+                index: 8,
+                argument: 0x1AA,
+                options: HardwareOptions::None,
+                kind: CommandKind::Control,
+                rsp_type: ResponseType::Short,
+                rsp_crc: true,
+                buffer: None,
+            })
+            .await?
+        {
+            Response::Short { value, .. } => {
+                tracing::trace!("CMD8 response: {value:#x}");
```

Comment:
take it or leave it, not a big deal: a similar trace is included in each of the methods that send a command to the card...what do you think about adding this to the `cmd` method, instead, like:
```rust
     tracing::trace!("CMD{index}: response {value:#}");
```
so that we don't have to repeat it in every other method?

Reply:
That would indeed be better. However, I guess you would have to match on the response in `cmd` then?
Or derive `Debug` on `Response`, which is probably the easiest solution.

Reply:
:woman_shrugging: if it's too much extra work, i'm fine leaving this as is!